### ISSUE TITLE
KSP NuGet Package

### ISSUE THREAD

**Issue by Pharylon** [RemoteTechnologiesGroup/RemoteTech#294]
I saw you broke dependency on those Tasks (awesome!) in 1.6. I thought I'd also let you know that there's now a KSP NuGet package. By using that, you could actually have the solution building as soon as it was cloned.

**Comment by Peppie23** [comment]
Any disadvantages? I'm not familiar with NuGet Packages :sweat_smile: help me to understand

**Comment by hakan42** [comment]
Where would I obtain this package from? Would make building on public Jenkins instances much easier :smile:

**Comment by Pharylon** [comment]
Right click on References, chose Manage NuGet Packages. In the left pane, make sure you're click on "Online" and then search for "KSP".

As for disadvantages... none I know of. I suppose if whoever is maintaining this package stops you'd have to go back to not using a NuGet package, but that's it.

**Comment by erendrake** [comment]
We might have to change a bit about the build in order to have the package restored on the build box. One thing to make sure we do it not check in the packages folder that stores the nuget cache. 

I actually work quite a bit with nuget at work so i dont mind taking this ticket.

**Comment by pjf** [comment]
> As for disadvantages... none I know of. I suppose if whoever is maintaining this package stops you'd have to go back to not using a NuGet package, but that's it.

The only disadvantage I can spot is if these are copyrigted squad assets which are being distributed, and squad decides to issue a take-down¹. But then yes, one goes back to not using a NuGet package.

¹ I've always presumed `Assembly-CSharp.dll` and friends were distribution-forbidden, but haven't checked. If they *can* be freely distributed that makes all my continuous integration tests much, much easier!

**Comment by Peppie23** [comment]
What about travis? Can travis handle the NuGet Packages? or has this only a benefit for contributors?

**Comment by erendrake** [comment]
tl;dr I dont think we should use the package.

@Peppie23 There should be no meaningful challenge to always using a nuget package. However i agree with @pjf that this is not a slamdunk because of how issued the package. 

If this was squad's package and they had signed up to keep it up to date. that would be fine and we would just use it. With it being from a 3rd party there are the issues of 

* how long will it take this random internet person to update their package
* how confident are we that this person hasnt mucked with the binaries
* is it legal for this person to distribute the dlls in this way. 

This would make life considerably easier for us but until is a squad official package I dont know if i can recommend that we use it.

**Comment by Peppie23** [comment]
i agree with you.

**Comment by Pharylon** [comment]
Yeah, that's probably true. I was thinking the legal part wouldn't matter, because it's not *your* legal problem. But it's true there's no promise that the NuGet package is the "real" version, so yeah, I suppose it's better not to use it.

There is one other way to compile-on-download, though.. make sure that Visual Studio can find the dll. You can manually give it multiple places to look if you want. To do so:

1. Right-click your csproj file and and edit it manually.

2. Find a line that looks like this: 

    <Reference Include="Assembly-CSharp">
      <HintPath>Assembly-CSharp.dll</HintPath>
      <Private>False</Private>
    </Reference>

3. Add more HintPaths. Now, by default, VS will only take the *last* one. So that should be your default. So... 

    <Reference Include="Assembly-CSharp">
      <HintPath Condition="Exists('C:\Program Files (x86)\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll')">C:\Program Files (x86)\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll</HintPath>	
      <HintPath Condition="Exists('C:\Program Files (x86)\Steam\SteamApps\common\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll')">C:\Program Files (x86)\Steam\SteamApps\common\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll</HintPath>
      <HintPath Condition="Exists('Assembly-CSharp.dll')">Assembly-CSharp.dll</HintPath>
      <Private>False</Private>
    </Reference>

Should make it look in the common places (since it takes the last HintPath, this would make it look in the root folder of the project, then the default steam install, then the default Squad install).

**Comment by Pharylon** [comment]
Yeah, that's probably true. I was thinking the legal part wouldn't matter, because it's not *your* legal problem. But it's true there's no promise that the NuGet package is the "real" version, so yeah, I suppose it's better not to use it.

There is one other way to compile-on-download, though.. make sure that Visual Studio can find the dll. You can manually give it multiple places to look if you want. To do so:

First, right-click your csproj file and and edit it manually and find a line that looks like this: 

    <Reference Include="Assembly-CSharp">
      <HintPath>Assembly-CSharp.dll</HintPath>
      <Private>False</Private>
    </Reference>

Then add more HintPaths. Now, by default, VS will only take the *last* one. So that should be your default. So... 

    <Reference Include="Assembly-CSharp">
      <HintPath Condition="Exists('C:\Program Files (x86)\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll')">C:\Program Files (x86)\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll</HintPath>	
      <HintPath Condition="Exists('C:\Program Files (x86)\Steam\SteamApps\common\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll')">C:\Program Files (x86)\Steam\SteamApps\common\Kerbal Space Program\KSP_Data\Managed\Assembly-CSharp.dll</HintPath>
      <HintPath Condition="Exists('Assembly-CSharp.dll')">Assembly-CSharp.dll</HintPath>
      <Private>False</Private>
    </Reference>

Should make it look in the common places (since it takes the last HintPath, this would make it look in the root folder of the project, then the default steam install, then the default Squad install).

**Comment by Pharylon** [comment]
P.S. This is how I have my fork set up. If you want me to send you a pull request for it, let me know.

**Comment by Peppie23** [comment]
tl;dr why do we need this? Sorry but maybe i don't understand this topic but every developer should know how to setup his own environment to compile this code.

When i don't have ksp on your HintPaths i become the same problem. The simplest way is to create a csproj.user (VisualStudio maybe Mono too) and change the Lib-Path like me, you also can add your own post build event etc...

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <Reference Include="Assembly-CSharp">
      <HintPath>D:\KerbalDev\assembly\Assembly-CSharp.dll</HintPath>
      <Private>false</Private>
    </Reference>
    <Reference Include="UnityEngine">
      <HintPath>D:\KerbalDev\assembly\UnityEngine.dll</HintPath>
      <Private>false</Private>
    </Reference>
  </ItemGroup>
  <PropertyGroup>
    <PostBuildEvent>
set KERBAL_VERSION="v0_90_0"
set GAME_DATA="$(OutDir)..\..\$(SolutionName)\"
set SAVE_PATH="$(OutDir)..\..\..\saves\"
set GAMEPATH="D:\KerbalDev\%25KERBAL_VERSION%25"
set DestPath="%25GAMEPATH%25\GameData\"
set DestPathSave="%25GAMEPATH%25\saves\"
set Binary="%25GAMEPATH%25\KSP_x86.exe"

:: Alten Dateien loeschen
del /S /F /Q "%25DestPath%25$(SolutionName)"
:: Neue rueber kopieren
xcopy "%25GAME_DATA%25*" "%25DestPath%25$(SolutionName)" /s /i
xcopy "%25SAVE_PATH%25*" "%25DestPathSave%25" /s /i
:: generierte dll loeschen
del "RemoteTech.dll"
:: exec!
%25Binary%25 -force-opengl
    </PostBuildEvent>
  </PropertyGroup>
</Project>
```
